pyomo/contrib/pynumero/src/hsl_interface/ma27Interface.c

```diff
@@ -0,0 +1,242 @@
+	return ma27->CNTL[i];
+}
+int get_info(struct MA27_struct* ma27, int i) {
+	return ma27->INFO[i];
+}
+
+// Functions for allocating WORK/FACT arrays:
+void alloc_iw_a(struct MA27_struct* ma27, int l) {
+	ma27->LIW_a = l;
+	ma27->IW_a = malloc(l*sizeof(int));
+	if (ma27->IW_a == NULL) { abort_bad_memory(1); }
+	ma27->IW_a_allocated = true;
+}
+void alloc_iw_b(struct MA27_struct* ma27, int l) {
+	ma27->LIW_b = l;
+	ma27->IW_b = malloc(l*sizeof(int));
+	if (ma27->IW_b == NULL) { abort_bad_memory(1); }
+	ma27->IW_b_allocated = true;
+}
+void alloc_a(struct MA27_struct* ma27, int l) {
+	ma27->LA = l;
+	//ma27->A = realloc(A, l*sizeof(double));
+	ma27->A = malloc(l*sizeof(double));
+	if (ma27->A == NULL) { abort_bad_memory(1); }
+	//memcpy(ma27->A, A, NZ*sizeof(double));
+	ma27->A_allocated = true;
+}
+
+void do_symbolic_factorization(struct MA27_struct* ma27, int N, int NZ, 
+		int* IRN, int* ICN) {
+
+	if (!ma27->IW_a_allocated) {
+		int min_size = 2*NZ + 3*N + 1;
+		int size = (int)(ma27->IW_factor*min_size);
+		alloc_iw_a(ma27, size);
+	}
+
+	ma27->IKEEP = malloc(3*N*sizeof(int));
+	if (ma27->IKEEP == NULL) { abort_bad_memory(1); }
+	ma27->IKEEP_allocated = true;
+	ma27->IW1 = malloc(2*N*sizeof(int));
+	if (ma27->IW1 == NULL) { abort_bad_memory(1); }
+
+	ma27ad_(&N, 
+			&NZ, 
+			IRN, 
+			ICN, 
+			ma27->IW_a,
+			&(ma27->LIW_a),
+			ma27->IKEEP, 
+			ma27->IW1, 
+			&(ma27->NSTEPS),
+			&(ma27->IFLAG),
+			ma27->ICNTL, 
+			ma27->CNTL,
+			ma27->INFO, 
+			&(ma27->OPS));
+
+	free(ma27->IW1);
+	free(ma27->IW_a);
+	ma27->IW_a_allocated = false;
+}
+
+void do_numeric_factorization(struct MA27_struct* ma27, int N, int NZ, 
+		int* IRN, int* ICN, double* A) {
+
+	// Get memory estimates from INFO, allocate A and IW
+	if (!ma27->A_allocated) {
+		int info5 = ma27->INFO[5-1];
+		int size = (int)(ma27->A_factor*info5);
+		alloc_a(ma27, size);
+		// A is now allocated
+	}
+	// Regardless of ma27->A's previous allocation status, copy values from A.
+	memcpy(ma27->A, A, NZ*sizeof(double));
+
+	if (!ma27->IW_b_allocated) {
+		int info6 = ma27->INFO[6-1];
+		int size = (int)(ma27->IW_factor*info6);
+		alloc_iw_b(ma27, size);
+	}
+
+	ma27->IW1 = malloc(N*sizeof(int));
+	if (ma27->IW1 == NULL) { abort_bad_memory(1); }
+
+	ma27bd_(&N, 
+			&NZ, 
+			IRN,
+			ICN,
+			ma27->A, 
+			&(ma27->LA),
+			ma27->IW_b, 
+			&(ma27->LIW_b),
+			ma27->IKEEP,
+			&(ma27->NSTEPS),
+			&(ma27->MAXFRT),
+			ma27->IW1,
+			ma27->ICNTL,
+			ma27->CNTL,
+			ma27->INFO);
+
+	free(ma27->IW1);
+}
+
+void do_backsolve(struct MA27_struct* ma27, int N, double* RHS) {
+
+	ma27->W = malloc(ma27->MAXFRT*sizeof(double));
+	if (ma27->W == NULL) { abort_bad_memory(1); }
+	ma27->IW1 = malloc(ma27->NSTEPS*sizeof(int));
+	if (ma27->IW1 == NULL) { abort_bad_memory(1); }
+  
+	ma27cd_(
+			&N, 
+			ma27->A, 
+			&(ma27->LA),
+			ma27->IW_b, 
+			&(ma27->LIW_b),
+			ma27->W,
+			&(ma27->MAXFRT),
+			RHS,
+			ma27->IW1,
+			&(ma27->NSTEPS), 
+			ma27->ICNTL, 
+			ma27->INFO
+			);
+
```

Comment:
Is there a particular reason not to have an iterative refinement subroutine here like the one you have for ma57?

Reply:
MA27 doesn't have a built-in iterative refinement subroutine, and I wanted the C code to just provide wrappers around the built-in routines. Iterative refinement can be implemented for MA27 on the Python side, but I haven't gotten around to this yet. (Haven't gotten around to writing a Python-side iterative refinement function for MA57 either, although this will hopefully be much simpler due to the built-in routine.)

Reply:
so m57 and mumps have that in the C side. It would be convenient to have the iterative refinement for ma27 in C too if possible. I implemented some subroutines for iterative refinement in python at some point and it was not very clean. It also add overhead given that the matrix vector products are fairly cheap

Reply:
I think doing the iterative refinement on the python side is a good approach. I'd like to keep the C code as simple as possible.

Reply:
Plus, we can always add it in later if needed. I think this is good for now.